pkg/ring/incremental_transfer.go

```diff
@@ -0,0 +1,252 @@
+
+	var lastErr error
+
+	for replica := 0; replica < rf; replica++ {
+		// Find the ranges which our token is holding data for as one of the replicas.
+		ends, err := d.Predecessors(token, replica, healthy)
+		if err != nil {
+			lastErr = err
+			continue
+		}
+
+		for _, end := range ends {
+			// Find the starting value for the range.
+			start, err := d.Neighbor(end.StatefulToken(), -1, false, healthy)
+			if err != nil {
+				lastErr = err
+				continue
+			}
+
+			// The target token's ingester is the ingester that either has data we
+			// want or should have data we'll send. When joining, it's the old
+			// end of the replica set. When leaving, it's the new end of the replica set.
+			target, err := d.Neighbor(start.StatefulToken(), rf, false, healthy)
+			if err != nil {
+				lastErr = err
+				continue
+			}
+
+			// If our local ingester is already in the replica set, we don't need to
+			// transfer. Our current token will never be considered since it's only
+			// ever LEAVING or JOINING at this point.
+			checkRange := RangeOptions{
+				Range: TokenRange{From: end.Token, To: target.Token},
+
+				ID:          i.ID,
+				IncludeFrom: true,
+				IncludeTo:   true,
+			}
+			if ok, _ := d.InRange(checkRange, healthy); ok {
+				continue
+			}
+
+			addr := d.Ingesters[target.Ingester].Addr
+			ret[addr] = append(ret[addr], TokenRange{From: start.Token, To: end.Token})
+		}
+	}
+
+	if lastErr != nil {
+		level.Error(util.Logger).Log("msg", fmt.Sprintf("failed to find complete transfer set for %d", token.Token), "err", lastErr)
+	}
+	return ret, lastErr == nil
+}
+
+// joinIncrementalTransfer will attempt to incrementally obtain chunks from
+// neighboring lifecyclers that contain data for token ranges they will
+// no longer receive writes for.
+func (i *Lifecycler) joinIncrementalTransfer(ctx context.Context) error {
+	// Make sure that we set all tokens to ACTIVE, even when we fail.
+	defer func() {
+		i.setTokens(i.getTransitioningTokens())
+		i.setTokensState(ACTIVE)
+		i.changeState(ctx, ACTIVE)
+	}()
+
+	r := i.getDenormalisedRing(ctx)
+
+	replicationFactor := i.cfg.RingConfig.ReplicationFactor
+	if active := r.FindIngestersByState(ACTIVE); len(active) < replicationFactor {
+		if len(active) == 0 {
+			return fmt.Errorf("no ingesters to request data from")
+		}
+
+		// Request an entire copy of all chunks from the first replica.
+		fullRange := []TokenRange{{0, math.MaxUint32}}
+		err := i.incTransferer.RequestChunkRanges(ctx, fullRange, active[0].Addr, false)
+		if err != nil {
+			level.Error(util.Logger).Log("msg", fmt.Sprintf("failed to request copy of data from %s", active[0].Addr))
+		}
+		return nil
+	}
+
+	var completesMtx sync.Mutex
+	pendingCompletes := make(map[string][]TokenRange)
+
+	for _, token := range i.transitioningTokens {
+		// Mark our token to JOINING and add it to our copy of the ring.
+		token.State = JOINING
+		r.RefreshTokenState(i.ID, append(i.getTokens(), token), false)
```

Comment:
Yes, we're only doing it so that `findTransferWorkload` detects the new token properly. Other ingesters don't need to know about the new token until we've accepted all data for the  new token.